docs/OBJECT-LEVEL-ENCRYPTION.md

```diff
@@ -0,0 +1,119 @@
+2. Zedagent receives the device configuration from the controller and parses cipher contexts and publishes them to the different agents using [types.CipherContext](../pkg/pillar/types/cipherinfotypes.go)
+3. Zedagent publishes the ECDH certificate received from TPM manager to the controller.
+
+### Downloader
+
+For downloading images from private data stores, downloader need to decrypt the data store credentials.
+
+Downloader passes the cipher context and controller certificate subscription contexts and cipher block from the datastore config to the cipher package for the decryption.
+
+The datastore credentials are only kept in memory in the downloader.
+
+### Network Interface Manager
+
+Nim needs to decrypt the WiFi credentials available in device port configuration for setting up the wlan connectivity. Nim calls the [device network](../pkg/pillar/devicenetwork) package for parsing the device port configuration.
+
+Nim passes device port configuration and subscription contexts of controller certificates and cipher contexts to the device network package. And, device network package passes received subscription contexts and cipher block from the WiFi configuration to the cipher package for the decryption.
+
+Decryption of WiFi credentials is very important if the device has only wlan connectivity, and depends on the key information being published persistently so that the ECDH calculation and decryption can be performed on reboot before the device can communication with the controller.
+
+The WiFi credentials are only kept in memory in nim, however in order to hand them to the wpa_supplicant process they are stored in /run which is a filesystem backed by memory.
+
+### Domain Manager
+
+Domain manager needs to decrypt the sensitive user data from the app instance configuration used for cloud-init and for container environment settings. Depending on the application instance such information can include local and remove keys and credentials.
+
+Domain manager passes the cipher block from the app configuration and subscription contexts of controller certificates and cipher context to the cipher package for the decryption.
+
+The user data are only kept in memory in the domainmgr, but for cloud-init it is used to create a CDROM image which is passed to the application. This CDROM image lives in /var/run/domainmgr/cloudinit which is a filesystem backed by memory.
+
+## Refreshing of the controller certificates in EVE
+
+Zedagent fetches the controller certificates at boot. So, we need a mechanism of refreshing the controller certificates in EVE if the controller certificates get updated when the device is already running.
+
+There are two ways for getting updated certificates:
+
+1. We can include hashes of the certificates on the top level of the device configuration itself. With the help of it, we can parse the certificates from the the device configuration in zedagent and will fetch the certificates if changed.
```

Comment:
This isn't clear. I think you are saying that the device configuration will include object encryption cert hash on each configuration, which will allow the receiver (zedagent?) to check if it has that hash. If not, it can re-request certificates from the endpoint. But I wasn't sure. In general, this paragraph should be more factual and assertive: 

"Include hash of the certificate in the device configuration. Upon receipt of a message, zedagent checks the hash against the hash of the controller certificate already stored. If it does not match, request the latest certificate from the controller." Something like that.

Reply:
This part has yet to be implemented, hence the two choices and the loose tone.

Reply:
Adding a "TBD" to this section to make that more clear.